### PR TITLE
refactor(launch): take build out of runner.run

### PR DESCRIPTION
**Pull request by bcsherma**
Fixes
-----
- Fixes WB-14335
- Fixes WB-13870

Description
-----------
**What does the PR do?**

This PR factors all build logic out of the runner class. The `launch` command or agent will instead call `builder.build` directly and then pass an image uri to the runner.

This PR also fixes an issue where `maybe_create_imagepull_secret` was not being called for image jobs.

This involves direct changes in all runners, so while (hopefully by time you read this) the unit and system tests are passing for the runners, it is a big refactor that touches a lot of stuff.

**Why did Ben do this?**

This paves the way for us to parallelize build and monitor it well. It also means we don't have the build logic implemented once per runner, which means build logic is much easier to fix or extend. For example, it was tricky to get macros enabled for each runner because the build logic changes from runner to runner. To fix issues there, this refactor was really helpful.

**Copilot**

### Summary
🚀🧹🐛

This pull request refactors and simplifies the code for the launch module and its runners, which are used to run wandb projects on different backends. It also updates the test code to use the new runner interface and a mock launch project. The main changes involve using the `image_uri` and `registry` parameters to run docker images, and removing unused or redundant code.

> _`Runner` classes change_
> _Simplify and refactor_
> _Autumn of old code_

### Walkthrough
*  Add registry argument to runner constructors and methods, and use image_uri argument to run docker images for launch projects ([link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-2b65556423a0400911187b8988badd8e6e829f5c4eeb0d1e30688b081fed6e37L59-R59), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-2b65556423a0400911187b8988badd8e6e829f5c4eeb0d1e30688b081fed6e37L67-R66), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-a1c99a9aafd44b94d56556ed6142736d658582e5f917e79b8bf60a0540b97df9L197-R197), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-75b70c365dac28700af3ff8747334d736d608078fc88317bf6c726d2231978b2L549-R550), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-75b70c365dac28700af3ff8747334d736d608078fc88317bf6c726d2231978b2L555-R565), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-957b0f5d138487d36a9c83a1c9914297cc79c42e7818b3453117ae3c4d59220cR165-R166), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-957b0f5d138487d36a9c83a1c9914297cc79c42e7818b3453117ae3c4d59220cL176-R187), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-f30f4bdb4783218d203be4d423d5a2879670da4d3ed206c2210c377993dcf78bR194), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-f30f4bdb4783218d203be4d423d5a2879670da4d3ed206c2210c377993dcf78bL216-R217), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-f30f4bdb4783218d203be4d423d5a2879670da4d3ed206c2210c377993dcf78bL231-R232), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-f30f4bdb4783218d203be4d423d5a2879670da4d3ed206c2210c377993dcf78bL242-R247), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-d198b874910290ed62fe55277f90293a60d9fb49695074ddbbcf335e465c49baL162-R165), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L356-R358), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86R372), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L441-R446), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L551-R552), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83L99-R102), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83L121-R122), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L112-R115), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L126-R134), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-971309ad05d166697209cdb39e7af4378445d88f459aadd134a5509f7fd70a62L87-R98))
* Remove unused imports, arguments, and variables from runner classes and methods, and simplify logic for constructing docker commands ([link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-d198b874910290ed62fe55277f90293a60d9fb49695074ddbbcf335e465c49baL18), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-d198b874910290ed62fe55277f90293a60d9fb49695074ddbbcf335e465c49baL26-R25), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-d198b874910290ed62fe55277f90293a60d9fb49695074ddbbcf335e465c49baL129-R132), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L11-L12), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L496-R498), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L506-R526), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L585-L590), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83L11-R12), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83L140-R159), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L11-L12), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7R16), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L137), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L203-R205), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7L228-R220), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-971309ad05d166697209cdb39e7af4378445d88f459aadd134a5509f7fd70a62L17-R20))
* Remove unnecessary creation of registry and builder for testing backends, and monkeypatch maybe_create_imagepull_secret function for kubernetes tests ([link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-0ccdfd1667c469c16b51091f502456a0a4d0c5814482765809e3b557ab60dceeL42-L43), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-0ccdfd1667c469c16b51091f502456a0a4d0c5814482765809e3b557ab60dceeL50-R56), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-0ccdfd1667c469c16b51091f502456a0a4d0c5814482765809e3b557ab60dceeL71-L72), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-0ccdfd1667c469c16b51091f502456a0a4d0c5814482765809e3b557ab60dceeR82), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-0ccdfd1667c469c16b51091f502456a0a4d0c5814482765809e3b557ab60dceeL102-R111), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-2b65556423a0400911187b8988badd8e6e829f5c4eeb0d1e30688b081fed6e37L51-L52))
* Add entrypoint and image_uri attributes to launch project class, and use them to build and run docker images for launch projects ([link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-b74194ae3f2489a233ad16701dc33678c9d4af7fc8c93023b554777e7095546dR8), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-b74194ae3f2489a233ad16701dc33678c9d4af7fc8c93023b554777e7095546dR14-R17), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-b74194ae3f2489a233ad16701dc33678c9d4af7fc8c93023b554777e7095546dL21-R26), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-75b70c365dac28700af3ff8747334d736d608078fc88317bf6c726d2231978b2L549-R550), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-75b70c365dac28700af3ff8747334d736d608078fc88317bf6c726d2231978b2L555-R565), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-957b0f5d138487d36a9c83a1c9914297cc79c42e7818b3453117ae3c4d59220cR165-R166), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-957b0f5d138487d36a9c83a1c9914297cc79c42e7818b3453117ae3c4d59220cR173), [link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86L648-R643))
* Add print statement for debugging launch config in `launch.py` ([link](https://github.com/wandb/wandb/pull/5825/files?diff=unified&w=0#diff-957b0f5d138487d36a9c83a1c9914297cc79c42e7818b3453117ae3c4d59220cR173))

### CODE REVIEW

**Comment by bcsherma** [comment]
🚨It's critically important that this is not merged without pull secret injection re-implemented!!!

File: wandb/sdk/launch/runner/kubernetes_runner.py [hunk 1]
```diff
@@ -504,36 +499,35 @@ def _inject_defaults(
...
+        #         "No registry specified. Please specify a registry in your wandb/settings file or pass a registry to the builder."
+        #     )
+        # secret = maybe_create_imagepull_secret(
+        #     core_api, builder.registry, launch_project.run_id, namespace
+        # )
+        # if secret is not None:
+        #     pod_spec["imagePullSecrets"] = [
+        #         {"name": f"regcred-{launch_project.run_id}"}
+        #     ]
+
```
Comment (bcsherma) [inline wandb/sdk/launch/runner/kubernetes_runner.py:530]: 🚨It's critically important that this is not merged without pull secret injection re-implemented!!!

File: wandb/sdk/launch/launch.py [hunk 2]
```diff
@@ -162,20 +162,29 @@ def _run(
...
     launch_project = fetch_and_validate_project(launch_project, api)
+    entrypoint = launch_project.get_single_entry_point()
+    image_uri = launch_project.docker_image  # Either set by user or None.
 
     # construct runner config.
     runner_config: Dict[str, Any] = {}
     runner_config[PROJECT_SYNCHRONOUS] = synchronous
 
     config = launch_config or {}
+    print(launch_config)
```
Comment (bcsherma) [inline wandb/sdk/launch/launch.py:173]:
```suggestion
```

File: wandb/sdk/launch/runner/kubernetes_runner.py [hunk 3]
```diff
@@ -493,36 +495,35 @@ def _inject_defaults(
...
-            if secret is not None:
-                pod_spec["imagePullSecrets"] = [
-                    {"name": f"regcred-{launch_project.run_id}"}
-                ]
             containers[0]["image"] = image_uri
             launch_project.fill_macros(image_uri)
 
+        if not self.registry:
+            raise LaunchError(
+                "No registry specified. Please specify a registry in your wandb/settings file or pass a registry to the builder."
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes_runner.py:517]: This is the wrong file name

**Review by KyleGoyette — COMMENTED** [review]
A few concerns and light fixes required

File: wandb/sdk/launch/runner/kubernetes_runner.py [hunk 4]
```diff
@@ -504,36 +506,35 @@ def _inject_defaults(
...
-                core_api, builder.registry, launch_project.run_id, namespace
-            )
-            if secret is not None:
-                pod_spec["imagePullSecrets"] = [
-                    {"name": f"regcred-{launch_project.run_id}"}
-                ]
             containers[0]["image"] = image_uri
             launch_project.fill_macros(image_uri)
 
+        if not self.registry:
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/runner/kubernetes_runner.py:515]: You are right!

File: wandb/sdk/launch/runner/abstract.py [hunk 5]
```diff
@@ -126,7 +125,11 @@ class AbstractRunner(ABC):
     (e.g. to run projects against your team's in-house cluster or job scheduler).
     """
 
-    def __init__(self, api: Api, backend_config: Dict[str, Any]) -> None:
+    def __init__(
+        self,
+        api: Api,
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/runner/abstract.py:127]: The AbstractRunner does not set __init__ as an abstractmethod. I can change that in this PR. I like the pattern that the sdk internals use where they have a super class implement function() and then have _function() as an abstractmethod and I think refactoring the builder/runner/registry to use this pattern would be a nice refactor.

File: wandb/sdk/launch/runner/local_container.py [hunk 6]
```diff
@@ -138,38 +138,28 @@ def run(
...
-                    image_uri,
-                    env_vars,
-                    docker_args=docker_args,
-                )
-            ).strip()
+            assert launch_project.docker_image == image_uri
+            pull_docker_image(image_uri)
+
+        additional_args = (
+            launch_project.override_args if launch_project.docker_image else None
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/runner/local_container.py:147]: I think that if we build the image we bake those overrides into the image's default entrypoint

[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/runner/local_container.py:147]: I don't think my PR changes this behavior but it is a little weird. What a user submits a run for which the image has already been built, BUT they want to set different override args? In that case I think launch_project.docker_image would be None

File: wandb/sdk/launch/agent/agent.py [hunk 7]
```diff
@@ -546,16 +546,23 @@ def _thread_run_job(
...
+            assert entrypoint is not None
+            image_uri = builder.build_image(project, entrypoint, job_tracker)
+        backend = loader.runner_from_config(
+            resource, api, backend_config, environment, registry
+        )
         _logger.info("Backend loaded...")
         api.ack_run_queue_item(job["runQueueItemId"], project.run_id)
-        run = backend.run(project, builder, job_tracker)
+        assert image_uri
+        run = backend.run(project, image_uri)
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/agent/agent.py:565]: Yes, do you think it's better if we pass the entrypoint to run? We are passing the launch project for a bunch of other reasons anyway

File: .circleci/config.yml [hunk 8]
```diff
@@ -1477,8 +1477,8 @@ workflows:
             - "functional-tests"
           matrix:
             parameters:
-              python_version_major: [ 3 ]
-              python_version_minor: [ 9 ]
+              python_version_major: [3]
```
Comment (KyleGoyette) [inline .circleci/config.yml:1480]: This files changes still need to be reverted

**Review by KyleGoyette — APPROVED** [review]

**Pull request merged**